src/mlpack/methods/ann/rbm/rbm.hpp

```diff
@@ -0,0 +1,447 @@
+ * @tparam PolicyType The RBM variant to be used (BinaryRBM or SpikeSlabRBM).
+ */
+template<
+  typename InitializationRuleType,
+  typename DataType = arma::mat,
+  typename PolicyType = BinaryRBM
+>
+class RBM
+{
+ public:
+  using NetworkType = RBM<InitializationRuleType, DataType, PolicyType>;
+  typedef typename DataType::elem_type ElemType;
+
+  /**
+   * Initialize all the parameters of the network using initializeRule.
+   *
+   * @tparam initializeRule InitializationRule object for
+   *         initializing the network parameter.
+   * @param predictors Training data to be used.
+   * @param visibleSize Number of visible neurons.
+   * @param hiddenSize Number of hidden neurons.
+   * @param batchSize Batch size to be used for training.
+   * @param numSteps Number of Gibbs Sampling steps.
+   * @param negSteps Number of negative samples to average negative gradient.
+   * @param persistence Indicates whether to use Persistent CD or not.
+   */
+  RBM(arma::Mat<ElemType> predictors,
+      InitializationRuleType initializeRule,
+      const size_t visibleSize,
+      const size_t hiddenSize,
+      const size_t batchSize = 1,
+      const size_t numSteps = 1,
+      const size_t negSteps = 1,
+      const size_t poolSize = 2,
+      const ElemType slabPenalty = 8,
+      const ElemType radius = 1,
+      const bool persistence = false);
+
+  // Reset the network.
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+  Reset();
+
+  // Reset the network.
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  Reset();
+
+  /**
+   * Train the feed-forward network on the given input data.
+   *
+   * This will use the existing model parameters as a starting point for the
+   * optimization. If this is not what you want, then you should access the
+   * parameters vector directly with Parameters() and modify it as desired.
+   *
+   * @param optimizer Optimizer type.
+   */
+  template<typename OptimizerType>
+  void Train(OptimizerType& optimizer);
+
+  /**
+   * Evaluate the RBM network with the given parameters.
+   * The function is needed for monitoring the progress of the network.
+   *
+   * @param parameters Matrix model parameters.
+   * @param i Index of the data point.
+   * @param batchSize Variable to store the present number of inputs.
+   */
+  double Evaluate(const arma::Mat<ElemType>& parameters,
+                  const size_t i,
+                  const size_t batchSize);
+
+  /**
+   * This function calculates the free energy of the BinaryRBM.
+   * The free energy is given by:
+   * $-b^Tv - \sum_{i=1}^M log(1 + e^{c_j+v^TW_j})$.
+   *
+   * @param input The visible neurons.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, double>::type
+  FreeEnergy(arma::Mat<ElemType>&& input);
```

Comment:
hm, do you think it would make sense to put this into the policy itself, we would have to pass the weights and the bias, but perhaps this would make the class cleaner?

Reply:
One of my concerns is with the inconvenience of using the policy paradigm and passing parameters, and then the number of duplicate matrices and objects that Kris' original code had.